proton/cert_pinning.py

```diff
@@ -0,0 +1,98 @@
+import base64
+import hashlib
+import json
+import os
+
+import requests
+from requests.adapters import HTTPAdapter
+from requests.packages.urllib3.connectionpool import HTTPSConnectionPool
+from requests.packages.urllib3.poolmanager import PoolManager
+
+crypto = requests.packages.urllib3.contrib.pyopenssl.OpenSSL.crypto
+
+class InspectedHTTPSConnectionPool(HTTPSConnectionPool):
+    """Custom HTTPSConnectionPool that verifies the certificate for each connection"""
+    
+    def _validate_conn(self, conn):
+        r = super(InspectedHTTPSConnectionPool, self)._validate_conn(conn)
+ 
+        sock = conn.sock
+        sock_connection = sock.connection
+
+        if self.is_session_secure(sock_connection.get_peer_cert_chain()[0], conn):
+            return r
+
+
+    def is_session_secure(self, cert, conn):
+        """Checks if connection is secure"""
+        
+        cert_hash = self.extract_hash(cert)
+
+        if not self.validate_hash(self.host, cert_hash):
+            # Also generate a report
```

Comment:
Where are you generating the report here?

Reply:
It is not generating it yet, but rather a reminder that it needs to be generated. My thought was to do that once most of the code is ready.